File: homework4/cpp/ring.c
#include <stdio.h>
#include <mpi.h>
#include <stdlib.h>


int main(int argv, char *argc){
    int ierr;
    int nNodes;
    int rank;
    int message = 0;
    int source,destination;

    ierr = MPI_Init(NULL, NULL);
    //get number of total processors
    MPI_Comm_size(MPI_COMM_WORLD, &nNodes);
    //get rank of processor
    MPI_Comm_rank(MPI_COMM_WORLD,&rank);

    if (rank == 0){
        message = 4;
    }

    for(source = 0; source < nNodes;source++){
        
        //make sure destination is the right one
        if(source == nNodes-1){
            destination = 0;
        }else{
            destination = source+1;
        }

        
        if(source == rank){
            MPI_Send(&message,1,MPI_INT, destination,0,MPI_COMM_WORLD);
            printf("processor %d sent the message: %d to processor %d\n", source, message, destination);

        }else if(destination == rank){
            MPI_Recv(&message,1,MPI_INT, source,0,MPI_COMM_WORLD,MPI_STATUS_IGNORE);
            printf("processor %d received the message %d from processor %d\n", destination, message, source);

        }    
    }

    ierr = MPI_Finalize();
    return ierr;
}

File: homework4/turnIn/ring/ring.f90
PROGRAM ring
INCLUDE 'mpif.h'

INTEGER :: ierr, rank,nNodes,i
INTEGER :: source = 0
INTEGER :: destination = 1
INTEGER :: message = 4
INTEGER :: stat(MPI_STATUS_SIZE)

!initialize MPI
CALL MPI_INIT(ierr);
!get number of processors
CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nNodes,ierr)
!get the rank of the processor
CALL MPI_COMM_RANK(MPI_COMM_WORLD,rank,ierr)

IF(rank .EQ. source) THEN
    message = 5
END IF

DO i = 0,nNodes-1
    source = i
    IF(source .EQ. nNodes-1) THEN
        destination = 0
    ELSE 
        destination = i+1
    END IF

    IF(rank .EQ. source) THEN
        CALL MPI_Send(message,1,MPI_INT,destination,0,MPI_COMM_WORLD,ierr)
        print *,"message sent:",message,"from processor", rank, "to processor", destination 
    ELSE IF(rank .EQ. destination) THEN
        CALL MPI_Recv(message,1,MPI_INT,source,0,MPI_COMM_WORLD,stat,ierr)
        print *,"message received:",message,"on processor", rank , "from processor", source
    END IF

END DO


CALL MPI_Finalize(ierr);
    
END PROGRAM ring
